sql/src/main/java/io/crate/jobs/PageDownstreamContext.java

```diff
@@ -96,30 +99,36 @@ private boolean isExhausted(int bucketIdx) {
     }
 
     public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultListener pageResultListener) {
-        synchronized (listeners) {
-            listeners.add(pageResultListener);
-        }
-        synchronized (lock) {
-            LOGGER.trace("setBucket: {}", bucketIdx);
-            if (allFuturesSet.get(bucketIdx)) {
-                pageDownstream.fail(new IllegalStateException("May not set the same bucket of a page more than once"));
-                return;
+        keepAliveFähnchen.set(KeepAliveFähnchen.ACTIVE);
+        try {
+            synchronized (listeners) {
+                listeners.add(pageResultListener);
             }
+            synchronized (lock) {
+                LOGGER.trace("setBucket: {}", bucketIdx);
+                if (allFuturesSet.get(bucketIdx)) {
+                    pageDownstream.fail(new IllegalStateException("May not set the same bucket of a page more than once"));
+                    return;
+                }
 
-            if (pageEmpty()) {
-                LOGGER.trace("calling nextPage");
-                pageDownstream.nextPage(new BucketPage(bucketFutures), new ResultListenerBridgingConsumeListener());
-            }
-            setExhaustedUpstreams();
+                if (pageEmpty()) {
+                    LOGGER.trace("calling nextPage");
+                    pageDownstream.nextPage(new BucketPage(bucketFutures), new ResultListenerBridgingConsumeListener());
+                }
+                setExhaustedUpstreams();
 
-            if (isLast) {
-                exhausted.set(bucketIdx);
-            }
-            bucketFutures.get(bucketIdx).set(rows);
-            allFuturesSet.set(bucketIdx);
+                if (isLast) {
+                    exhausted.set(bucketIdx);
+                }
+                bucketFutures.get(bucketIdx).set(rows);
+                allFuturesSet.set(bucketIdx);
 
-            clearPageIfFull();
+                clearPageIfFull();
+            }
+        } finally {
+            keepAliveFähnchen.set(KeepAliveFähnchen.PASSIVE);
```

Comment:
the bucketFutures could have listeners attached which are executed in another thread - so this isn't necessarily synchronous.
I think the set call could be moved into the PageConsumeListener